// needle-sdk/core/classes.js
class RequestData {
    constructor() {
        this.data = [];
        this.remote_addr = '';
        this.request_method = '';
        this.http_host = '';
        this.path_info = '';
        this.http_user_agent = '';
        this.incident_action = '';
        this.incident_module = '';
        this.resp_obj = null;
        this.req_start_ts = null;
        this.perf_data = [];
    }
}

class InstrMethod {
    constructor(sec_module, module, orig_method, is_instr) {
        this.sec_module = sec_module;
        this.module = module;
        this.orig_method = orig_method;
        this.is_instr = is_instr;
    }
}

class PerfMarker {
    constructor(sec_module, lang_module, type, ts) {
        this.sec_module = sec_module;
        this.lang_module = lang_module;
        this.type = type;
        this.ts = ts;
    }
}

class PerfTime {
    constructor(sec_module, lang_module, time) {
        this.sec_module = sec_module;
        this.lang_module = lang_module;
        this.time = time;
    }
}

class PerfRecord {
    constructor(path, ts, perf_times) {
        this.path = path;
        this.req_ts = ts;
        this.perf_times = perf_times;
    }
}

module.exports = {
    RequestData: RequestData,
    InstrMethod: InstrMethod,
    PerfMarker: PerfMarker,
    PerfTime: PerfTime,
    PerfRecord: PerfRecord
};
